pkg/client/tools.go

```diff
@@ -123,11 +164,108 @@ func ImageImportIntoClusterMulti(ctx context.Context, runtime runtimes.Runtime,
+		if err != nil {
+			return fmt.Errorf("could not open image stream for given images %s: %w", imagesFromRuntime, err)
+		}
+		err = loadImageFromStream(ctx, runtime, stream, cluster)
+		if err != nil {
+			return fmt.Errorf("could not load image to cluster from stream %s: %w", imagesFromRuntime, err)
+		}
+		// load the images directly into the nodes
 
+	}
+
+	if len(imagesFromTar) > 0 {
+		// copy tarfiles to shared volume
+		l.Log().Infof("Importing images from %d tarball(s)...", len(imagesFromTar))
+
+		files := make([]*os.File, len(imagesFromTar))
+		readers := make([]io.Reader, len(imagesFromTar))
+		failedFiles := 0
+		for i, fileName := range imagesFromTar {
+			file, err := os.Open(fileName)
+			if err != nil {
+				l.Log().Errorf("failed to read file '%s', skipping. Error below:\n%+v", fileName, err)
+				failedFiles++
+				continue
+			}
+			files[i] = file
+			readers[i] = file
+		}
+		multiReader := io.MultiReader(readers...)
+		err := loadImageFromStream(ctx, runtime, io.NopCloser(multiReader), cluster)
+		if err != nil {
+			return fmt.Errorf("could not load image to cluster from stream %s: %w", imagesFromTar, err)
+		}
+		for _, file := range files {
+			err := file.Close()
+			if err != nil {
+				l.Log().Errorf("Failed to close file '%s' after reading. Error below:\n%+v", file.Name(), err)
+			}
+		}
+	}
 	return nil
+}
 
+func loadImageFromStream(ctx context.Context, runtime runtimes.Runtime, stream io.ReadCloser, cluster *k3d.Cluster) error {
+	var errorGroup errgroup.Group
+
+	numNodes := 0
+	for _, node := range cluster.Nodes {
+		// only import image in server and agent nodes (i.e. ignoring auxiliary nodes like the server loadbalancer)
+		if node.Role == k3d.ServerRole || node.Role == k3d.AgentRole {
+			numNodes++
+		}
+	}
+	// multiplex the stream so we can write to multiple nodes
+	pipeReaders := make([]*io.PipeReader, numNodes)
+	pipeWriters := make([]io.Writer, numNodes)
+	for i := 0; i < numNodes; i++ {
+		reader, writer := io.Pipe()
+		pipeReaders[i] = reader
+		pipeWriters[i] = writer
+	}
+
+	errorGroup.Go(func() error {
+		_, err := io.Copy(io.MultiWriter(pipeWriters...), stream)
+		if err != nil {
+			return fmt.Errorf("failed to copy read stream. %v", err)
+		}
+		err = stream.Close()
+		if err != nil {
+			return fmt.Errorf("failed to close stream. %v", err)
+		}
+		return nil
+	})
+
+	pipeId := 0
+	for _, node := range cluster.Nodes {
+		// only import image in server and agent nodes (i.e. ignoring auxiliary nodes like the server loadbalancer)
+		if node.Role == k3d.ServerRole || node.Role == k3d.AgentRole {
+			pipeReader := pipeReaders[pipeId]
+			errorGroup.Go(func() error {
+				l.Log().Infof("Importing images into node '%s'...", node.Name)
+				if err := runtime.ExecInNodeWithStdin(ctx, node, []string{"ctr", "image", "import", "-"}, pipeReader); err != nil {
+					return fmt.Errorf("failed to import images in node '%s': %v", node.Name, err)
+				}
+				return nil
+			})
+			pipeId++
+		}
+	}
```

Comment:
Need to capture the node as well, otherwise it will use server-0 all the time and thus break (i.e. run forever), if the cluster has multiple nodes (see CI Run, e.g. https://drone-pr.rancher.io/rancher/k3d/519/1/4)
```suggestion
	for _, n := range cluster.Nodes {
		node := n
		// only import image in server and agent nodes (i.e. ignoring auxiliary nodes like the server loadbalancer)
		if node.Role == k3d.ServerRole || node.Role == k3d.AgentRole {
			pipeReader := pipeReaders[pipeId]
			errorGroup.Go(func() error {
				l.Log().Infof("Importing images into node '%s'...", node.Name)
				if err := runtime.ExecInNodeWithStdin(ctx, node, []string{"ctr", "image", "import", "-"}, pipeReader); err != nil {
					return fmt.Errorf("failed to import images in node '%s': %v", node.Name, err)
				}
				return nil
			})
			pipeId++
		}
	}
```